### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by dbcesar**
Depends on: 
https://github.com/rock-perception/perception-orogen-apriltags/pull/11

### CODE REVIEW

File: tasks/Task.cpp [hunk 1]
```diff
@@ -107,14 +111,13 @@ void Task::updateHook()
...
 
     //read input ports
-    std::vector<base::Vector2d> corners;
+    std::vector<apriltags::VisualFeaturePoint> corners_vector;
     base::LinearAngular6DCommand setpoint;
-    bool no_corners = (_marker_corners.read(corners) == RTT::NoData);
-    bool no_setpoint = (_cmd_in.read(setpoint) == RTT::NoData);
-    ctrl_state.timestamp = base::Time::now();
+    bool no_corners = (_marker_corners.read(corners_vector) != RTT::NewData);
+    bool no_setpoint = (_cmd_in.read(setpoint) != RTT::NewData);
```
Comment (doudou) [inline tasks/Task.cpp]: Why do you require a new setpoint each time ?

Apart from breaking general conventions (unless in a nested control loop, setpoints are either sent only once, or at low frequency), it creates a race condition between the corners and setpoint: you could read the setpoint in one updateHook and the corner in another, leading to not doing anything.

File: tasks/Task.cpp [hunk 2]
```diff
@@ -129,28 +132,40 @@ void Task::updateHook()
 
     if (state() == CONTROLLING)
```
Comment (doudou) [inline tasks/Task.cpp]: It's really hard to follow long branches. So, don't do

```
if (X)
{
  VERY LONG PIECE OF CODE
}
else
{
  VERY SMALL PIECE OF CODE
}
<END OF FUNCTION>
```

Instead do

```
if (! X) {
  SMALL PIECE OF CODE
  return;
}

LONG PIECE OF CODE
<END OF FUNCTION>
```

File: tasks/Task.cpp [hunk 3]
```diff
@@ -283,14 +300,37 @@ void Task::setGain()
...
+    visp::saturationValues saturation = _saturation.get();
+
+    for (int i=0; i < 3; ++i)
+    {
+        // write NaN on the non-desirable outputs in order to
+        // keep the consistence with the other controllers
+        // linear
+        if (!expected_inputs.linear[i])
+        {
+            v[i] = base::NaN<double>();
```
Comment (doudou) [inline tasks/Task.cpp]: Is there really a reason why you first update the `v` array and only then write in `vel` ? You could directly write `vel` ... (which would arguably be more readable ...)

File: tasks/Task.cpp [hunk 4]
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
...
+    //since the non-expected input are NaN, set it to default values
+    //in order to have valid values for the HomogeneusMatrix creation
+    for (int i = 0; i < 3; ++i)
+    {
+        if (!expected_inputs.linear[i]) {cmd_in_body.linear[i] = 0;}
+        if (!expected_inputs.angular[i])
+        {
+            if ( i == 0 || i == 1)
+            {
+                cmd_in_body.angular[i] = -M_PI_2;
```
Comment (doudou) [inline tasks/Task.cpp]: You really would need to explain that one with a comment.

File: tasks/Task.cpp [hunk 5]
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
...
+    vpTranslationVector bdto(cmd_in_body.linear[0], cmd_in_body.linear[1], cmd_in_body.linear[2]);
+    vpRxyzVector bdro(cmd_in_body.angular[0], cmd_in_body.angular[1], cmd_in_body.angular[2]);
+    vpRotationMatrix bdRo(bdro);
+    //pose of the object wrt the body
+    vpHomogeneousMatrix bdMo(bdto, bdRo);
+
+    //bdMo is the desired pose of the object in the body frame
+    //cdMo is the desired pose of the object in the camera frame
+    //cMb is the transformation matrix of the body in the camera frame
+    cdMo = cMb * bdMo;
```
Comment (doudou) [inline tasks/Task.cpp]: There's really no good reason to store this as an attribute. The only place where you use it is just after you called transformInput. Make it the return value of the method and use it as a local variable.

File: tasks/Task.cpp [hunk 6]
```diff
@@ -301,13 +341,135 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
...
+}
+
+vpMatrix Task::getJacobianFromExpectedInputs(visp::expectedInputs expected_inputs)
+{
+    vpMatrix eJe;
+    eJe.eye(6);
+
+    for(int i = 0; i < 3; ++i)
+    {
+        if (!expected_inputs.linear[i])  {eJe[i][i] = 0;}
```
Comment (doudou) [inline tasks/Task.cpp]: This is a really weird style.

If you want to have it a bit compressed, simply use the syntax without `{}`

```
if (bla)
   blo;
```

Avoid one-liner, it makes code coverage difficult to understand.

File: tasks/Task.cpp (hunk 1 above)
Comment (dbcesar) [inline tasks/Task.cpp]: I think now it is better

File: tasks/Task.cpp (hunk 2 above)
Comment (dbcesar) [inline tasks/Task.cpp]: Done

File: tasks/Task.cpp (hunk 3 above)
Comment (dbcesar) [inline tasks/Task.cpp]: No, there is not.  Changed.

File: tasks/Task.cpp (hunk 4 above)
Comment (dbcesar) [inline tasks/Task.cpp]: Done

File: tasks/Task.cpp (hunk 6 above)
Comment (dbcesar) [inline tasks/Task.cpp]: Changed here. Is there a better choice between

```
if  (bla)
    blo;
```

or 

```
if (bla)
{
    blo;
}

```

?

File: tasks/Task.cpp (hunk 5 above)
Comment (dbcesar) [inline tasks/Task.cpp]: Acctualy, I also use it on the line 264.

```
//////////////////////////////
// Update Feature 3 - t, tu //
//////////////////////////////
cdMc = cdMo*cMo.inverse();
t.buildFrom(cdMc);
tu.buildFrom(cdMc) 
```

File: tasks/Task.cpp (hunk 4 above)
Comment (doudou) [inline tasks/Task.cpp]: **Formatting** If you have special cases for two cases out of three .... Either use a separate default_angular_values array that you use in the loop. Or don't use a loop at all. Ifs are really hindering understanding.

**However** .... This `-M_PI_2` according to your comment is basically applying the frame-to-body transformation. Which you apply further down. So I still don't get it.

File: tasks/Task.cpp [hunk 7]
```diff
@@ -301,13 +337,144 @@ base::samples::RigidBodyState Task::convertToRbs(vpHomogeneousMatrix pose)
...
 }
+
+void Task::transformInput(base::LinearAngular6DCommand cmd_in_body)
+{
+    //initialize setpoint on the body frame
+    //since the non-expected input are NaN, set it to default values
+    //in order to have valid values for the HomogeneusMatrix creation
+    for (int i = 0; i < 3; ++i)
+    {
+        if (!expected_inputs.linear[i]) {cmd_in_body.linear[i] = 0;}
```
Comment (doudou) [inline tasks/Task.cpp]: Single line ifs should be formatted

```
if (X)
  do_y;
```

Single-line stuff is harmful in general, as it hinders code coverage measurement (there are two separate branches on the same line)

Generally speaking, be consistent with yourself and with the rest of the codebase. This is the only place where you format a single-line if like this. In all other cases you use

```
if (X)
{
   do_y
}
```

File: tasks/Task.cpp [hunk 8]
```diff
@@ -283,14 +299,34 @@ void Task::setGain()
...
+    for (int i=0; i < 3; ++i)
+    {
+        // write NaN on the non-desirable outputs in order to
+        // keep the consistence with the other controllers
+        // linear
+        if (!expected_inputs.linear[i])
+        {
+            vel.linear[i] = base::NaN<double>();
+        }
+        else if (v[i] > saturation.linear[i])
```
Comment (doudou) [inline tasks/Task.cpp]: what about negative saturations ???

File: tasks/Task.cpp (hunk 5 above)
Comment (doudou) [inline tasks/Task.cpp]: This method is part of the same code flow. Return the matrix from both transformInput (which should really be called computeDesiredPose) and then from updateDesiredPose (hell, maybe there's something wrong with this naming scheme ... IMO you should just merge both methods together, the latter really just "finishes" the former's job). Then, pass it as argument.

Generally speaking, avoid storing as attribute anything that does not need to be long-lived. And avoid side-effects in methods as long as it does not complexify the code base too much.

A method that computes a single value (like this one) should really be side-effect free (i.e. const) and return its result. Let the caller store it somewhere if it really has to (in this case, it does not). Doing so improves readability by making the data flow obvious (it would be obvious in updateHook that you compute the new desired pose and then compute the command from the new desired pose).

Finally, drop the boolean return values. All of them (except the ones from the `*Hook` methods that are mandated by RTT, obviously). Use exceptions if there are errors. You need to learn the reflex of asking yourself what was the point of having `bool` return values everywhere if **you never test against them**

File: tasks/Task.cpp (hunk 8 above)
Comment (dbcesar) [inline tasks/Task.cpp]: Indeed. Fixed it.

File: tasks/Task.cpp (hunk 4 above)
Comment (dbcesar) [inline tasks/Task.cpp]: >  However .... This -M_PI_2 according to your comment is basically applying the frame-to-body transformation. Which you apply further down. So I still don't get it.

The -M_PI_2 indeed is related to the frame-to-body transformation, however the loop is just to assing real values to the ones that are NaN (since they are not expected inputs). This is done in order to allow the Homogeneus Matrix creation.

File: tasks/Task.cpp (hunk 5 above)
Comment (dbcesar) [inline tasks/Task.cpp]: I merged the methods and now its way better.

I also tried to eliminate unecessary attributes, but still not sure if I can get rid of someone else. 
I tried also to turn the methods as much side effect free as possible, let me now if I can improve something else.

Concerning the bool methods...I took it out and now throw exeptions when it goes to an undesired behaviour. However I think I should handle  it somehow. Would you recomnend the creation of one more runtime state (for instance :WAITING_VALID_TARGET) and move the component there anytime these exception are thrown?

File: tasks/Task.cpp [hunk 9]
```diff
@@ -31,25 +30,31 @@ bool Task::configureHook()
...
+    target_list = _target_list.get();
+    desired_target = _desired_target.get();
+
     // Initialize camera parameters.
     // The library desconsider the tangencial values, that's why only the
-    // radial values are informed. 
+    // radial values are informed.
     frame_helper::CameraCalibration cal = _camera_parameters.get();
+    double scaling = _scaling.get();
+    if(scaling != 1.0)
```
Comment (doudou) [inline tasks/Task.cpp]: Don't. There's no harm in multiplying by one, get rid of the if.

File: tasks/Task.cpp [hunk 10]
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
         state(CONTROLLING);
     }
 
-    if (state() == CONTROLLING)
+    if (!(state() == CONTROLLING))
```
Comment (doudou) [inline tasks/Task.cpp]: This is `state() != X`

File: tasks/Task.cpp [hunk 11]
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
...
+        writeVelocities(v);
+        return;
+    }
 
-        // Update this jacobian in the task structure. It will be used to compute
-        // the velocity skew (as an articular velocity)
-        // qdot = -lambda * L^+ * cVe * eJe * (s-s*)
-        task.set_eJe(eJe) ;
+    // filter the corners of an desired target
+    Points corners;
```
Comment (doudou) [inline tasks/Task.cpp]: Point corners = value; instead of `Point corners; corners = value;`

File: tasks/Task.cpp [hunk 12]
```diff
@@ -127,30 +129,46 @@ void Task::updateHook()
...
+    corners = filterTarget(corners_vector, desired_target);
 
-        // Set the gain
-        setGain();
+    // transforms the setpoint from the body frame to the camera frame and
+    // build the Homogeneus Matrix of the desired object pose in the camera
+    // frame
+    vpHomogeneousMatrix cdMo;
```
Comment (doudou) [inline tasks/Task.cpp]: Same. Initialize inline instead of doing `init; assignation;`

File: tasks/Task.cpp [hunk 13]
```diff
@@ -237,77 +258,225 @@ void Task::updateFeatures(std::vector<base::Vector2d> corners)
     p.set_xyZ(x, y, Z);
     depth.buildFrom(P.get_x(), P.get_y(), Z, log(Z/Zd));
 
-    //////////////////////////////
-    // Update Feature 3 - t, tu //
-    //////////////////////////////
+    /**
+     *  Update Feature 3 - t, tu
+     */
     cdMc = cdMo*cMo.inverse();
```
Comment (doudou) [inline tasks/Task.cpp]: Any point in having this one stored as attribute ?

File: tasks/Task.cpp (hunk 6 above)
Comment (doudou) [inline tasks/Task.cpp]: I personally prefer the first.

File: tasks/Task.cpp (hunk 5 above)
Comment (doudou) [inline tasks/Task.cpp]: > WAITING_VALID_TARGET

You can wait for a target. You should complain loudly (i.e. raise an exception **and terminate**) if you get an **invalid** target.

File: tasks/Task.cpp (hunk 4 above)
Comment (doudou) [inline tasks/Task.cpp]: Still don't get it.

The normal workflow should be

input_values -> normalized_input_values -> transformation -> homogeneous matrix.

I.e. replace the NaN by values that make sense in _the input value space_ and then apply the transformation.

Having two different code paths for the NaN and normal cases spells trouble. In general, the less code paths there is, the better.

Comment (dbcesar) [inline tasks/Task.cpp]: Not sure If I understood, because I think I'm following the suggested workflow:

> input_values -> normalized_input_values -> transformation -> homogeneous matrix.

_Input values_ : the argument "setpoint"-> 
_normalized_input_values_: the "_for_" loop -> 
_transformation_: building the homogeneus matrix from the setpoint (bdMo)and multiplying by the transformation matrix (cMb) -> 
_homogeneus matrix_: the result of the transformation.

Am I missing something?

> Having two different code paths for the NaN and normal cases spells trouble. In general, the less code paths there is, the better.

I don't see how to have only a single path in this case. Any suggestions?

Comment (doudou) [inline tasks/Task.cpp]: Let's assume that I'm the one missing something for now ...
- in what frame is the setpoint expressed ?

Comment (dbcesar) [inline tasks/Task.cpp]: The _setpoint_ is the object pose wrt body frame .

However,  visp library works with the object pose in the _camera_ frame and that is why I use the body-to-camera matrix to transform it.

**Comment by dbcesar** [comment]
@doudou I've added the documentation explain the conventions of this component. Especially the cmd_in (setpoint) case. 

I've also changed the way of find and update the target parameters. I've tried to follow your suggestion of return the vector index, but handle the cases on the updateHook (on the states checking) became more complex. So I've created two methods to read the input ports. 
In the case of the corners vector reading, it returns a boolean if either there is no new data on the input port or the desired target is not detected.

**Comment by dbcesar** [comment]
@doudou ping

**Comment by dbcesar** [comment]
ping @doudou

**Comment by doudou** [comment]
:+1: Sorry for the delay in answering

**Comment by dbcesar** [comment]
Sorry, but I couldn't track your answering.

**Comment by dbcesar** [comment]
Sorry, but I couldn't track your answering.

**Comment by dbcesar** [comment]
Any new comments here?

**Comment by doudou** [comment]
Nope. That was what the thumbs-up meant ;-)
